[NIFI-987] Added Processor For Writing Events to Riemann

rickysaltzer opened the pull request:
- Introduced nifi-riemann-bundle for future Riemann backed monitoring

- Added initial PutRiemann processor for writing events to Riemann
  using the Riemann batch client.

	- Values for events are provided using the NiFi expression language
		e.g. Metric -> ${latency.milliseconds:divide(1000)}

joewitt commented:
Ricky,

Sorry for the delay in providing feedback.  Can you look into why the branch has conflicts?  Perhaps I'm misunderstanding what I'm seeing from github.  Also, the PR doesn't appear to address the necessary changes to the nifi-assembly/NOTICE and provide one for the nar itself so that we ensure the EPL 1.0 dependency is properly annotated.  I haven't checked what dependencies it pulls in yet but wanted to put that out there first.

Thanks
Joe

rickysaltzer commented:
Hey Joe -

Not a problem on the delay! I'll take a look when I get a chance. Thanks!

Thanks,
Ricky

rickysaltzer commented:
The conflict was inside of the `pom.xml` inside the `nifi-nar-bundles` module. The new couchbase module was conflicting with the new riemann module inside the `<modules>` stanza. I made sure both modules are in there, and rebased the branch on the current upstream master.

busbey commented:
Please edit commit message to start with JIRA key and brief summary. i.e.:

```commit
NIFI-987 Add Reimann monitoring

- Introduced nifi-riemann-bundle for future Riemann backed monitoring

- Added initial PutRiemann processor for writing events to Riemann
  using the Riemann batch client.

	- Values for events are provided using the NiFi expression language
		e.g. Metric -> ${latency.milliseconds:divide(1000)}
```

busbey commented:
Why do this as a processor rather than as a reporting task?

trkurc commented:
just speculation, but if you have a bunch of metrics in a file from somewhere and want to post it to Riemann, a processor would be useful. Granted, when I think of people using Riemann, I think of a bunch of distributed processes all sending out live metrics/monitoring information.

rickysaltzer commented:
@busbey good question. I envision that we can later extend the Riemann bundle to allow internal NiFi reporting to Riemann, similar to how Ambari was recently added. For now, there are plenty of situations where it would be useful to allow reporting of a flowfile's existence and/or attributes to Riemann for basic data flow reporting and monitoring. 

Let me give you a few examples:

* Send a notification to Riemann for each flowfile passed through a specific processor.  This not only allows for monitoring, but also alerting if desired. 
* Each node in the NiFi cluster can periodically send a message with a TTL that if expires would indicate that the node in the cluster is no longer running. 
* Immediate alerting when specific data flow scenarios are met (e.g. failed to extract archive, failed to hit an API). 
* To extend on what @trkurc said, you could very easily report metrics from files if desired. Or - you could use the ExecuteProcess to collect metrics using arbitrary commands. 

I'm sure there are other use cases that can be applied.

busbey commented:
that all sounds reasonable to me, thanks for the explanation!

rickysaltzer commented:
@bbende I updated the review with a different approach to sending events. After I started doing a lot of random disconnect testing, it appeared that the `RiemannBatchClient` had some really odd behavior when it came to detecting a down connection. Instead of using that, I'm using the regular `RiemannClient` and batching them together myself. This way I can call the `deref()` on the `IPromise` in order to detect connection failures, or timeouts. 

If the connection drops while sending events, the processor will `yield()` and attempt to re-connect on the next trigger until it can do so successfully. I also added a test (`testFailedDeref`) to verify that this works, as well as tested it internally with both `TCP` and `UDP`.

rickysaltzer commented on nifi-nar-bundles/nifi-riemann-bundle/pom.xml:
```diff
@@ -0,0 +1,53 @@
+  <dependencies>
+    <dependency>
+      <groupId>com.aphyr</groupId>
+      <artifactId>riemann-java-client</artifactId>
```
this is the only outside dependency, which is [Apache Licensed](https://github.com/aphyr/riemann-java-client/blob/master/LICENSE)

joewitt replied:
Ricky,

Here is a flow chart for properly handling licensing.

1) Check license of dependency.  If it is clear what you should do go to step 2.
2) You are done.  But first, was it easy?  If yes go back to step 1 until it is painful.  Once painful go to step 3.
3) Done.

So first question is can we find this dependency without adding a special maven repo to default maven configs?  If not then it has problematic ripple effect of telling folks to pull from additional repos.  

Second question, once you dig into that riemann client codebase you see this pom https://github.com/aphyr/riemann-java-client/blob/master/riemann-java-client/pom.xml.  That shows us there are several additional compile time deps which means transitively we'll have those too.  So you need to validate their licenses and their transitive deps licenses.  If needed will need to update license/notice/etc..

Thanks
Joe

busbey replied:
one eas(ier) way to check the transitive dependencies is to use "mvn dependency:list" within your module. that will give you a set of group/artifact/versions you can use to grab individual licenses.

since we build the modules with the resource plugin, you can just go to the target/ directory and unpack oyur jar file. within, there will be a file `META-INF/DEPENDENCIES` that will list what all the transitive dependency poms claim the license is. Usually they'll include the library's website if you want to verify.

olegz commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+        builder.setHost(host.getValue());
+      }
+      PropertyValue tags = context.getProperty(ATTR_TAGS).evaluateAttributeExpressions(flowFile);
+      if (tags.getValue() != null && !tags.getValue().equals("")) {
```
Kind of general comment applying to all IF statements above. Since you are checking for empty strings, should/would values such as "  " be still considered an empty string? If so, should it be trimmed before checking for it being empty?

rickysaltzer replied:
I believe so, I can add that check

busbey commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+        builder.setService(service.getValue());
+      }
+      PropertyValue description = context.getProperty(ATTR_DESCRIPTION).evaluateAttributeExpressions(flowFile);
+      if (description.getValue() != null && !description.getValue().trim().equals("")) {
```
if you're willing to add a dependency on Apache Commons-Lang, these checks can be simplified via [StringUtils.isBlank](https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#isBlank(java.lang.String))

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+    .description("Port that Riemann is listening on")
+    .required(true)
+    .defaultValue("5555")
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Non-empty validator is redundant with the integer validator provided below and the fact that this value is also required

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+    .required(true)
+    .defaultValue("5555")
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .addValidator(StandardValidators.INTEGER_VALIDATOR)
```
StandardValidators.POSITIVE_INTEGER_VALIDATOR is more appropriate for a port

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+    .required(false)
+    .defaultValue("100")
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .addValidator(StandardValidators.INTEGER_VALIDATOR)
```
Ditto on these validators being redundant

apiri commented:
You are not including your nar in the nifi-assembly pom (and would want to include this in the root pom.xml as a managed dependency for versioning) and it was not getting included in the generated assemblies.

Additionally, it seems as though your nifi-riemann-processors does not have the requisite META-INF/services/org.apache.nifi.processor.Processor file for SPI.  Accordingly, it was not getting loaded on NiFi startup as an available processor.

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+    .required(true)
+    .allowableValues(new Transport[]{Transport.TCP, Transport.UDP})
+    .defaultValue("TCP")
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Validator not needed given your use of allowable values

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+            client = RiemannClient.udp(host, port);
+            break;
+        }
+        client.connect();
```
I believe this is the culprit in what works out to be a resource leak.  When testing against my sample instance, and the server went down (running in a docker container), I quickly ended up in an OOME.  

```
2015-11-04 11:31:19,111 ERROR [Timer-Driven Process Thread-1] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] Failed to process session due to org.apache.nifi.processor.exception.ProcessException: Unable to connect to Riemann [192.168.99.100:5555] (TCP)
Connection failed: org.apache.nifi.processor.exception.ProcessException: Unable to connect to Riemann [192.168.99.100:5555] (TCP)
Connection failed
2015-11-04 11:31:20,114 ERROR [Timer-Driven Process Thread-5] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] failed to process due to java.lang.OutOfMemoryError: unable to create new native thread; rolling back session: java.lang.OutOfMemoryError: unable to create new native thread
2015-11-04 11:31:20,114 ERROR [Timer-Driven Process Thread-5] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] failed to process session due to java.lang.OutOfMemoryError: unable to create new native thread: java.lang.OutOfMemoryError: unable to create new native thread
2015-11-04 11:31:20,114 WARN [Timer-Driven Process Thread-5] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] Processor Administratively Yielded for 1 sec due to processing failure
2015-11-04 11:31:20,115 WARN [Timer-Driven Process Thread-5] o.a.n.c.t.ContinuallyRunProcessorTask Administratively Yielding PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] due to uncaught Exception: java.lang.OutOfMemoryError: unable to create new native thread
2015-11-04 11:31:20,119 WARN [Timer-Driven Process Thread-5] o.a.n.c.t.ContinuallyRunProcessorTask 
java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method) [na:1.8.0_60]
	at java.lang.Thread.start(Thread.java:714) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:950) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1368) [na:1.8.0_60]
	at org.jboss.netty.util.internal.DeadLockProofWorker.start(DeadLockProofWorker.java:38) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.openSelector(AbstractNioSelector.java:344) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.<init>(AbstractNioSelector.java:96) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.<init>(AbstractNioWorker.java:51) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorker.<init>(NioWorker.java:45) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.createWorker(NioWorkerPool.java:45) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.createWorker(NioWorkerPool.java:28) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorkerPool.newWorker(AbstractNioWorkerPool.java:99) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorkerPool.init(AbstractNioWorkerPool.java:69) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.<init>(NioWorkerPool.java:39) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.<init>(NioWorkerPool.java:33) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioClientSocketChannelFactory.<init>(NioClientSocketChannelFactory.java:152) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioClientSocketChannelFactory.<init>(NioClientSocketChannelFactory.java:117) ~[na:na]
	at com.aphyr.riemann.client.TcpTransport.connect(TcpTransport.java:151) ~[na:na]
	at com.aphyr.riemann.client.RiemannClient.connect(RiemannClient.java:155) ~[na:na]
	at org.apache.nifi.processors.riemann.PutRiemann.onScheduled(PutRiemann.java:260) ~[na:na]
	at org.apache.nifi.processors.riemann.PutRiemann.onTrigger(PutRiemann.java:286) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1134) ~[nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:127) [nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:49) [nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:119) [nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_60]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_60]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_60]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_60]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
```

Performing a cursory look through the code, it seems like when that IOException occurs the associated client should also be closed

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+        Proto.Msg returnMessage = riemannClient.sendEvents(eventsQueue).deref(writeTimeout, TimeUnit.MILLISECONDS);
+        if (returnMessage == null) {
+          context.yield();
+          throw new ProcessException("Timed out writing to Riemann!");
```
What are your thoughts on routing these to failure or perhaps even another relationship in terms of making communication?  Would give users more flexibility in handling differing circumstances in lieu of just kicking it back to source connection.

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+      riemannClient.flush();
+    } catch (Exception e) {
+      context.yield();
+      throw new ProcessException("Failed writing to Riemann\n" + e.getMessage());
```
Same commentary as other ProcessException

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+      throw new ProcessException("Failed writing to Riemann\n" + e.getMessage());
+    }
+    session.transfer(successfulFlowFiles, REL_SUCCESS);
+    session.commit();
```
Framework provides this at the conclusion of the onTrigger method as per AbstractProcessor#onTrigger(final ProcessContext context, final ProcessSessionFactory sessionFactory)

apiri commented on nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java:
```diff
@@ -0,0 +1,376 @@
+    if (riemannClient == null || !riemannClient.isConnected()) {
+      // clean up the client and attempt to re-initialize the processor
+      cleanUpClient();
+      onScheduled(context);
```
This block is what is spurring the leak when comms cannot be established to the remote system.  Is RiemannClient threadsafe?  This would be shared across multiple threads should concurrency which could cause issues given the logic currently in place.  If not, would likely prefer to see client generation per onTrigger.  Otherwise, need some better handling of the resource instantiation and teardown.

rickysaltzer commented:
Thanks for all of the comments, I'll address them as soon as I get a chance. As @apiri pointed out, If the connection fails, then we need to call `client.close()` (if not null) in order for the Riemann library to clean up its netty threads. Otherwise, it endlessly spawns threads which eventually causes the OOM exception. 

I don't particularly like the way I'm backing off when failing to connect, as it seems to retry very quickly. This is why the OOM surfaces minutes after killing Riemann. What is the best practice here for yielding for a longer period of time?  

        catch (IOException e) {
          if (client != null) {
            client.close();
          }
          context.yield();
          throw new ProcessException(String.format("Unable to connect to Riemann [%s:%d] (%s)\n%s", host, port, transport, e.getMessage()));
         }

apiri commented:
That context.yield is what is controlled by the processor configuration so that the user can choose how often this retry should happen, which I think is the right fit.  However, it could be interesting to have different yield strategies as options; as an example, an exponential backoff among them.

rickysaltzer commented:
@apiri let me know if there's anything else glaring I should address before we discuss committing it. If it helps, I've been running this NAR in production for about a month without any issues.

apiri commented:
@rickysaltzer Sounds good, have this on my list to scope out soon as time permits.  Thanks!

apiri commented:
@rickysaltzer Did another review on this and overall things look pretty good and the issues of bigger concern seem to be corrected.  I left an additional patch on the ticket with some stylistic items and version update (0.4.2-SNAPSHOT) changes I would like to apply on top of your patch.  

My final comments are concerning some of the properties.  In the case of the metric property, this is something that you use as a float to provide in the transmission to the Riemann server.  It would be good to include this in the property descriptor to make this a little clearer.  This should also be applied to the TTL and time properties.  I did some digging to try to figure out how to flesh this out more, but wasn't quite sure how to best describe them.  If you can provide some additional detail on those and the anticipated input, it would certainly help.  Apologies if this is a misunderstanding on my behalf, still quite the Riemann novice.

Thanks again for the updates though and with some additional details on the properties I think we are good to go!  Thanks for all your efforts!

rickysaltzer commented:
Thanks @apiri, I'll try and update a few of the descriptions to make things more clear and upload a new patch. Cheers

apiri commented:
Awesome, thanks!

rickysaltzer commented:
The latest patch should address most of the description concerns. I'm in the process of running all NiFi tests now and doing a final workflow test with a current `master` build. I will report results shortly. Thanks again.

rickysaltzer commented:
Tests pass, and my few workflows I created behave as expected. If it's okay with you @apiri, I'd like to re-base my changes throughout this PR into one commit before I push upstream.

apiri commented:
@rickysaltzer Sounds fair.  I had squashed/pushed to my personal fork at a35be5e358108a48b34de3fcf5edcb428a794d03 to make the review a little easier.  Additionally, there were a couple of style items I patched up as well as marking OnScheduled/OnStopped variables as volatile in 1d4fdfe57c9d966c0be6fd5cc537623854f4491c.  If they seem fair, I can merge them to accompany your contribution and push it all together.

Let me know what works and sounds good to you.  Thanks!

rickysaltzer commented:
Oh nice, I'll see about merging your changes after I squash mine. Would it be okay if I push this patch once it's ready? I haven't yet verified I'm even able to push upstream, and now seems like a good time to try that out.

apiri commented:
Yep, of course and please do!  Just was trying to save some unnecessary keystrokes if not needed.  Let me know if you hit any hurdles.  Presumably you can access https://github.com/apiri/incubator-nifi/commit/1d4fdfe57c9d966c0be6fd5cc537623854f4491c.patch to grab a patch for that commit and overlay it on yours.  I am sure there is a more elegant way, however I am also sure I don't know what that is off the top of my head ;)

Regardless, +1 with those changes and the work in your latest push to the PR.

rickysaltzer commented:
@apiri awesome, your patch applied without issue!
